### PR TITLE
Added detect port

### PR DESCRIPTION
**Pull request by xudafeng**
like title

### CODE REVIEW

**Comment by zhuyali** [comment]
https://github.com/vuejs/vue/issues/4452

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,4 @@
 node_modules
 .DS_Store
 docs/_book
+*.sw*
```
Comment (LinusBorg) [inline .gitignore]: I don't think this is relevant to this PR?

File: template/.gitignore [hunk 2]
```diff
@@ -9,3 +9,4 @@ test/unit/coverage
 test/e2e/reports
 selenium-debug.log
 {{/e2e}}
+*.sw*
```
Comment (LinusBorg) [inline template/.gitignore]: I don't think this is relevant to this PR?

Comment (xudafeng) [inline template/.gitignore]: @LinusBorg removed.

**Comment by LinusBorg** [comment]
Thanks! :) Will take this for a testride tomorrow, then probably merge it.

**Comment by xudafeng** [comment]
Thx team!

**Comment by yyx990803** [comment]
Hey thanks for the PR - however, the port is needed in e2e tests and need to be determined before running the tests. The async-ness of `detect-port` also makes `dev-server.js` no longer export the server instance. I think it's probably ok for the user to ensure the port is available.

**Pull request closed without merging**